src/sentry/shared_integrations/client/base.py

```diff
@@ -329,3 +378,66 @@ def get_with_pagination(
             if num_results < page_size:
                 return output
         return output
+
+    def record_response(self, response: BaseApiResponse):
+        if self.is_response_fatal(response):
+            self.record_request_fatal(response)
+        elif self.is_response_error(response):
+            self.record_request_error(response)
+        elif self.is_response_success(response):
+            self.record_request_success(response)
+
+    def record_error(self, error: Exception):
+        redis_key = self._get_redis_key()
+        if not len(redis_key):
+            return
+        if not self.is_error(error):
+            return
+        buffer = IntegrationRequestBuffer(redis_key)
+        buffer.record_error()
+        if buffer.is_integration_broken():
+            self.disable_integration()
+
+    def record_request_error(self, resp: Response):
+        redis_key = self._get_redis_key()
+        if not len(redis_key):
+            return
+        buffer = IntegrationRequestBuffer(redis_key)
+        buffer.record_error()
+        if buffer.is_integration_broken():
+            self.disable_integration()
+
+    def record_request_success(self, resp: Response):
+        redis_key = self._get_redis_key()
+        if not len(redis_key):
+            return
+        buffer = IntegrationRequestBuffer(redis_key)
+        buffer.record_success()
+
+    def record_request_fatal(self, resp: Response):
+        redis_key = self._get_redis_key()
+        if not len(redis_key):
+            return
+        buffer = IntegrationRequestBuffer(redis_key)
+        buffer.record_fatal()
+        if buffer.is_integration_broken():
+            self.disable_integration()
+
+    def disable_integration(self) -> None:
+        rpc_integration, rpc_org_integration = integration_service.get_organization_contexts(
+            integration_id=self.integration_id
+        )
+        oi = OrganizationIntegration.objects.filter(integration_id=self.integration_id)[0]
+        org = Organization.objects.get(id=oi.organization_id)
+        if features.has("organizations:disable-on-broken", org):
+            integration_service.update_integration(
+                integration_id=rpc_integration.id, status=ObjectStatus.DISABLED
+            )
+        self.logger.info(
+            "integration.disabled",
+            extra={
+                "integration_id": self.integration_id,
+                "provider": rpc_integration.provider,
+                "organization_id": rpc_org_integration[0].organization_id,
```

Comment:
@chloeho7 Possible `rpc_org_integration` list is empty. In that case, this would error out. We still probably should disable those integrations though, just without an email since there is no one to email.